<h1><b>Restricted Boltzmann Machine και Deep Belief Network</b></h1>
<p align="justify">Στην συγκεκριμένη άσκηση θα μελετήσετε τον τρόπο λειτουργίας μιας <i>RBM (<a href="https://en.wikipedia.org/wiki/Restricted_Boltzmann_machine">Restricted Boltzmann Machine</a>)</i> καθώς και των <i>DBN (<a href="https://en.wikipedia.org/wiki/Deep_belief_network">Deep Belief Network</a>)</i>, χρησιμοποιώντας το έτοιμο πρόγραμμα που σας δίνεται.Το συγκεκριμένο πρόγραμμα αξιοποιεί το <a href="https://en.wikipedia.org/wiki/MNIST_database">dataset του <i>MNIST</i></a>, όπου είναι μια μεγάλη βάση δεδομένων με χειρόγραφα ψηφία που χρησιμοποιείται συνήθως για την εκπαίδευση διαφόρων συστημάτων επεξεργασίας εικόνας. Για την άσκηση, θα πρέπει να χρησιμοποιήσετε το αρχείο <i>mnist_original.mat</i>, το οποίο είναι διαθέσιμο από <a href="https://www.kaggle.com/datasets/avnishnish/mnist-original?resource=download">εδώ</a>.</p>
<p align="justify">Μία αρκετά σημαντική εφαρμογή της <i>RBM</i> είναι η εξαγωγή χαρακτηριστικών (feature representation) από ένα dataset με σκοπό την αναπαράσταση της εισόδου (ορατοί νευρώνες) με ένα διάνυσμα μικρότερης διάστασης (κρυφοί νευρώνες). Στη συγκεκριμένη άσκηση θα συγκρίνετε την ακρίβεια ενός ταξινομητή ψηφίων με τη χρήση του αλγορίθμου <i>Logistic Regression</i>, όταν εκείνος δέχεται ως είσοδο το dataset (i) χωρίς να έχει υποστεί επεξεργασία από το <i>RBM</i>, (ii) αφου υποστεί επεξεργασία από το <i>RBM</i>, (iii) με τη χρήση <i>DBN</i>, δηλαδή δύο stacked <i>RBM</i>.</p>
<p align="justify"> Με βάση τον κώδικα που σας έχει δοθεί, καλείστε να απαντήσετε στα παρακάτω ερωτήματα:</p>
<ul>
<li>Να περιγράψετε σύντομα τον τρόπο λειτουργίας μιας <i>RBM</i>. Τι διαφορές έχει σε σχέση με μία <i> Μηχανή Boltzmann</i>;</li>
<li>Ποια είναι η λογική των <i>DBN</i> και σε τι προβλήματα τα αξιοποιούμε;</li>
<li>Να αναφέρετε τις βασικότερες εφαρμογές των <i>RBM</i> και <i>DBN</i>.</li>
<li>Εκτός από <i>RBM</i>, τι άλλα μοντέλα μπορούν να χρησιμοποιηθούν για να δημιουργήσουν <i>DBN</i>.</li>
<li>Συγκρίνετε τα αποτελέσματα της ταξινόμησης με τον αλγόριθμo <i>Logistic Regression</i> χωρίς τη χρήση <i>RBM</i> σε σχέση με τα αποτελέσματα της ταξινόμησης που έχει χρησιμοποιηθεί η <i>RBM</i> καθώς και με αυτή όπου χρησιμοποιούνται <i>RBM</i> και <i>DBN</i> για την εξαγωγή των χαρακτηριστικών. Τι παρατηρείτε ως προς την ακρίβεια των αποτελεσμάτων;</li>
</ul>

ΑΠΑΝΤΗΣΕΙΣ:


i) Η Restricted Boltzman Machine είναι ένα μοντέλο μη κατευθυνόμενου γράφου που υπάγεται στους αλγορίθμους βαθιάς μάθησης. Παίζει ένα σημαντικό ρόλο στη μείωση διαστάσεων , ταξινόμηση και παλλινδρόμηση. Η RBM είναι ένα βασικό κομμάτι του Deep-Belief network. Έχει 2 στρώματα νευρωνικών δικτύων. Το 1ο στρώμα της RBM καλείται visiblle layer (ορατό στρώμα) ή input layer (στρώμα εισόδου), ενώ το 2ο στρώμα είναι hidden layer (κρυμμένο στρώμα).
Στην RBM, οι διασυνδέσεις μεταξύ ορατών κόμβων και κρυμμένων κόμβων δημιουργούνται με τη χρήση συμμετρικών βαρών. Συγκεκριμένα, όλα οι hidden nodes συνδέονται με  όλους τους visible nodes, αλλά δεν υπάρχει σύνδεση μεταξύ κόμβων ίδιου στρώματος. Δηλαδή, δεν μπορούν να συνδεθούν μεταξύ τους 2 κόμβοι από το hidden layer, όπως επίσης δεν μπορούν να συνδεθούν μεταξύ τους κόμβοι από το visible layer.
Κάθε κόμβος δέχεται και επεξεργαζεται την είσοδο και παίρνει στοχαστικές αποφάσεις σχετικά με το αν θα μεταδώσει το μηνύμα ή όχι.
Ο αλγόριθμος αυτός ανήκει στη κατηγορία unsupervised learning, καθώς μέσα από τη διαδικασία να πηγαίνει μπρος και πίσω μεταξύ του ορατού (visible) και του κρυμμένου (hidden) στρώματος, η RBM ανακατασκευάζει τα δεδομένα.
Ο αλγόριθμος Restricted Boltzman Machine λέγεται Restricted εξαιτίας του περιορισμού σύνδεσης μεταξύ νευρώνων ίδιου στρώματος. Ο αλγόριθμος RBM βασίζεται σε έναν άλλον αλγόριθμο βαθιάς μάθησης που λέγεται Boltzman Machine.
Η κύρια διαφορά μεταξύ Restricted Boltzman Machine και Boltzman Machine είναι ότι στη Boltzman Machine, οι νευρώνες(nodes) δεν συνδέονται μόνο με τους νευρώνες(nodes) των άλλων στρωμάτων(layers), αλλά συνδέονται και μεταξύ τους. Δηλαδή, συνδέονται νευρώνες που βρίσκονται στο ίδιο layer. Είναι απαραίτητο κάθε νευρώνας του συστήματος να συνδέεται με όλους τους νευρώνες του συστήματος.Άρα, σε αντίθεση με τη Restricted Boltzman Machine, η Boltzman Machine έχει ασυμμετρία στα βάρη.
Επίσης, όταν η είσοδος αυξάνεται επ΄αόριστον, το feature δεν έχει αποδοτική επιρροή στη δομή. Για να αντιμετωπίσουμε αυτή τη κατάσταση δημιουργήθηκε ο αλγόριθμος Restricted Boltzman Machine. Η RBM προσπαθεί να εκπαιδεύσει το δίκτυο να αντιγράφει την είσοδο πιθανολογικά και να μαθαίνει τη πιθανότητα κατανομής της εισόδου.



ii) Τα Deep Belief Networks (DBNs) είναι νευρωνικά δίκτυα με πολλαπλά Restricted Boltzman Machine layers. Δηλαδή. με πολλά στρώματα RBM.
Επίσης, στα DBNs , η είσοδος του επόμενου στρώματος προέρχεται από την έξοδο του προηγούμενου στρώματος. Τα DBNs υιοθετούν μια ιεραρχικά μη επιβλεπόμενη άπληστη μέθοδο προεκπαίδευσης για να προεκπαιδεύσει ιεραρχικά κάθε Restricted Boltzman Machine.
Τα DBNs αναπαριστούν πιο περίπλοκα και αφηρημένα δεδομένα. Συγκεκριμένα, επειδή τα DBNs είναι επί των πλήστων παραγωγικά μοντέλα, αυτό που κάνουν είναι ότι βελτιώνουν την αναπαράσταση του feature με σκοπό να προσεγγίσουμε τη target function.
Με τη χρήση των DBNs, απλά features που παρέχονται στο 1ο στρώμα τοποθετούνται σε πιο περίπλοκες και αφηρημένες αναπαραστάσεις των δεδομένων. Χρησιμοποιούμε DBNs για να αναλύσουμε και να ταυτοποιήσουμε αντικείμενα στον πραγματικό κόσμο.
Τα RBM μπορούν να χρησιμοποιηθούν για την κατασκευή και την εκαίδευση των DBN που έχουν μεγαλύτερη αναπαραστατική ισχύ από ένα μόνο RBM. Για να εκπαιδεύσουμε ένα DBN, το εκπαιδεύουμε στρώμα στρώμα , ξεκινώντας από το πρώτο κρυφό στρώμα. Μόλις εκπαιδευτεί ένα συγκεκριμένο στρώμα, τα βάρη και τα biases(μεροληψίες) παγώνουν  για αυτό το στρώμα  και για το επόμενο κρυφό στρώμα πάνω από αυτό που εκπαιδεύονται. Οι τιμές που προκύπτουν από το προηγούμενο κρυφό στρώμα χρησιμοποιούνται ως δεδομένα για να εκπαιδεύσουν το επόμενο κρυφό στρώμα. Οι τιμές αυτές προκύπτουν από την είσοδο της εικόνας σε ένα visible layer. Η διαδικασία επαναλαμβάνεται μέχρι να εκπαιδευτεί και το τελευταίο στρώμα.

iii) Εφαρμογές RBM και DBN:
Αναγνώριση προτύπων σε εικόνες, ήχο , κείμενο και άλλα είδη δεδομένων.
Ταξινόμηση δεδομένων.
Συσταδοποίηση: Ομαδοποίηση δεδομένων σε συστάδες βάσει των ομοιοτήτων τους.
Μείωση διαστάσεων, επιλέγοντας εξόδους μικρότερου μεγέθους σε σχέση με την είσοδο με την είσοδο, δηλαδή το σύνολο των δειγματικών σημείων.

Εφαρμογή RBM - Αναπαραγωγή Χειρόγραφων Αριθμών :
Προβλήματα γενίκευσης από υπεραπλούστευση διαδικασίας μάθησης:
Λανθασμένη αναπαραγωγή χειρόγραφου αριθμού 3 από RBM με δείγμα
μάθησης αποκλειστικά με χειρόγραφα στοιχεία αριθμού 2

Δείγμα Μάθησης RBM με Ενσωμάτωση Label σε Εικόνες
Δείγμα Μάθησης Εικόνων με προσθήκη metadata: Κωδικοποίηση κλάσης σαν label
και ενσωμάτωση σε δειγματικά στοιχεία μάθησης που κλειδώνονται στην αρχική
κατάσταση των ορατών νευρώνων της RBM
Ταξινόμηση Δείγματος Test
Είσοδος εικόνας test χωρίς label και αναπαραγωγή της στην τελική κατάσταση των
ορατών νευρώνων της RBM με συμπλήρωση πληροφορίας κλάσης σύμφωνα με τις
στατιστικές εκτιμήσεις που προέκυψαν από τη διαδικασία μάθησης

iv) Άλλα μοντέλα που μπορούν να χρησιμοποιηθούν για τη δημιουργία DBN είναι:
Autoencoders: Οι Autoencoders είναι νευρωνικά δίκτυα που χρησιμοποιούνται για αυτόματη εκμάθηση χαρακτηριστικών. Αποτελούνται από έναν κωδικοποιητή (encoder) που μετατρέπει την είσοδο σε μια πιο συμπαγή αναπαράσταση (συνήθως μικρότερη διάσταση) και έναν αποκωδικοποιητή (decoder) που αποκωδικοποιεί την αναπαράσταση πίσω στην αρχική της μορφή. Μπορούν να χρησιμοποιηθούν ως βάση για τη δημιουργία DBN.

Convolutional Neural Networks (CNNs): Οι Convolutional Neural Networks είναι κατάλληλες για την επεξεργασία εικόνων και χαρακτηρίζονται από τη χρήση φίλτρων συνέλιξης που μπορούν να εντοπίσουν χαρακτηριστικά σε διαφορετικά επίπεδα πυκνότητας. Μπορούν επίσης να χρησιμοποιηθούν ως προεπεξεργαστές για την εξαγωγή χαρακτηριστικών πριν από τη χρήση ενός DBN.

Stacked Denoising Autoencoders (SDAEs): Οι Stacked Denoising Autoencoders είναι μια εξέλιξη των απλών Autoencoders, στην οποία κάθε στρώμα του δικτύου εκπαιδεύεται να απομακρύνει θόρυβο από την είσοδο. Η διαδικασία αυτή μπορεί να βοηθήσει στην εκπαίδευση πιο σταθερών και γενικευμένων χαρακτηριστικών, κάτι που είναι σημαντικό για τη δημιουργία DBN.

Generative Adversarial Networks (GANs): Οι Generative Adversarial Networks είναι μοντέλα που αποτελούνται από δύο ανταγωνιστικά δίκτυα, έναν γεννήτορα και έναν διακριτή, τα οποία εκπαιδεύονται αμοιβαία. Οι GANs μπορούν να χρησιμοποιηθούν για τη δημιουργία νέων δεδομένων με βάση μια δεδομένη κατανομή πιθανοτήτων και μπορεί να χρησιμοποιηθούν επίσης ως στρώματα προεπεξεργασίας για DBN.

In [ ]:
#!/usr/bin/env python
# source: https://devdreamz.com/question/905929-stacking-rbms-to-create-deep-belief-network-in-sklearn

import numpy as np
import matplotlib.pyplot as plt

from scipy.io import loadmat
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import BernoulliRBM
from sklearn.base import clone
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report

def norm(arr):
    arr = arr.astype(float)
    arr -= arr.min()
    arr /= arr.max()
    return arr

if __name__ == '__main__':

    # load MNIST data set
    mnist = loadmat("mnist-original.mat")
    X, Y = mnist["data"].T, mnist["label"][0]

    # normalize inputs to 0-1 range
    X = norm(X)

    # split into train, validation, and test data sets
    X_train, X_test, Y_train, Y_test = train_test_split(X,       Y,       test_size=10000, random_state=0)
    X_train, X_val,  Y_train, Y_val  = train_test_split(X_train, Y_train, test_size=10000, random_state=0)

    # --------------------------------------------------------------------------------
    # set hyperparameters

    learning_rate = 0.02
    total_units   =  800
    total_epochs  =   50
    batch_size    =  128

    C = 100. # optimum for benchmark model according to sklearn docs: https://scikit-learn.org/stable/auto_examples/neural_networks/plot_rbm_logistic_classification.html#sphx-glr-auto-examples-neural-networks-plot-rbm-logistic-classification-py)

    # --------------------------------------------------------------------------------
    # construct models

    # RBM
    rbm = BernoulliRBM(n_components=total_units, learning_rate=learning_rate, batch_size=batch_size, n_iter=total_epochs, verbose=1)

    # "output layer"
    logistic = LogisticRegression(C=C, solver='lbfgs', multi_class='multinomial', max_iter=200, verbose=1)

    models = []
    models.append(Pipeline(steps=[('logistic', clone(logistic))]))                                              # base model / benchmark
    models.append(Pipeline(steps=[('rbm1', clone(rbm)), ('logistic', clone(logistic))]))                        # single RBM
    models.append(Pipeline(steps=[('rbm1', clone(rbm)), ('rbm2', clone(rbm)), ('logistic', clone(logistic))]))  # RBM stack / DBN

    # --------------------------------------------------------------------------------
    # train and evaluate models

    for model in models:
        # train
        model.fit(X_train, Y_train)

        # evaluate using validation set
        print("Model performance:\n%s\n" % (
            classification_report(Y_val, model.predict(X_val))))

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Model performance:
              precision    recall  f1-score   support

         0.0       0.95      0.96      0.96       995
         1.0       0.96      0.98      0.97      1121
         2.0       0.90      0.90      0.90      1015
         3.0       0.90      0.88      0.89      1033
         4.0       0.93      0.92      0.92       976
         5.0       0.90      0.88      0.89       884
         6.0       0.94      0.94      0.94       999
         7.0       0.92      0.93      0.92      1034
         8.0       0.88      0.86      0.87       923
         9.0       0.89      0.90      0.89      1020

    accuracy                           0.92     10000
   macro avg       0.92      0.92      0.92     10000
weighted avg       0.92      0.92      0.92     10000


[BernoulliRBM] Iteration 1, pseudo-likelihood = -139.68, time = 14.82s
[BernoulliRBM] Iteration 2, pseudo-likelihood = -119.41, time = 20.84s
[BernoulliRBM] Iteration 3, pseudo-likelihood = -105.27, time = 20.60s
[Bernoul

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Model performance:
              precision    recall  f1-score   support

         0.0       0.98      0.98      0.98       995
         1.0       0.99      0.99      0.99      1121
         2.0       0.96      0.97      0.97      1015
         3.0       0.97      0.97      0.97      1033
         4.0       0.98      0.97      0.97       976
         5.0       0.98      0.96      0.97       884
         6.0       0.99      0.99      0.99       999
         7.0       0.97      0.98      0.97      1034
         8.0       0.96      0.96      0.96       923
         9.0       0.96      0.96      0.96      1020

    accuracy                           0.97     10000
   macro avg       0.97      0.97      0.97     10000
weighted avg       0.97      0.97      0.97     10000


[BernoulliRBM] Iteration 1, pseudo-likelihood = -139.47, time = 15.44s
[BernoulliRBM] Iteration 2, pseudo-likelihood = -119.80, time = 21.87s
[BernoulliRBM] Iteration 3, pseudo-likelihood = -105.25, time = 25.74s
[Bernoul

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Model performance:
              precision    recall  f1-score   support

         0.0       0.98      0.99      0.99       995
         1.0       0.99      0.99      0.99      1121
         2.0       0.97      0.98      0.98      1015
         3.0       0.97      0.97      0.97      1033
         4.0       0.98      0.97      0.98       976
         5.0       0.97      0.96      0.97       884
         6.0       0.99      0.98      0.98       999
         7.0       0.98      0.98      0.98      1034
         8.0       0.96      0.96      0.96       923
         9.0       0.96      0.97      0.96      1020

    accuracy                           0.98     10000
   macro avg       0.98      0.98      0.98     10000
weighted avg       0.98      0.98      0.98     10000




v) Σε αυτό το σημείο θα αναλύσουμε τα αποτελέσματα που μας έδωσαν οι υλοποιήσεις των αλγορίθμων:

Το μοντέλο της λογιστικής παλινδρόμησης χωρίς RBM	 δίνει accuracy 92%.
Το μοντέλο της λογιστικής παλινδρόμησης με RBM δίνει accuracy 97%.
Το μοντέλο της λογιστικής παλινδρόμησης με DBN από 2 RBM δίνει accuracy	98%.

Επίσης, σε κάθε μοντέλο καταγράφεται το precision, το recall και το f1-score. Σε κάθε περίπτωση οι τιμές αυτές είναι πάνω από 90%, το οποίο δείχνει  μεγάλη αξιοπιστία και ακρίβεια όσο αφορά τις προβλέψεις, καθώς και μεγάλη εγκυρότητα για το σύνολο δεδομένων  που χρησιμοποιήθηκε.

Με βάση το αποτέλεσμα, τη κάλυτερη επίδοση σημειώνει το μοντέλο λογιστικής παλινδρόμησης με DBN από 2 RBM. Σίγουρα, η επίδραση που έχει η προσθήκη του/των RBM(s) στο συνολικό χρόνο εκτέλεσης δεν είναι αμελητέα, όμως η αύξηση της συνολικής ακρίβειας στην οποία οδηγούν σίγουρα έχει πολύ υψηλότερη βαρύτητα.

